src/blame.ts

```diff
@@ -0,0 +1,82 @@
+import gql from 'tagged-template-noop'
+import sourcegraph from 'sourcegraph'
+
+/**
+ * Resolve a URI of the forms git://github.com/owner/repo?rev#path and file:///path to an absolute reference, using
+ * the given base (root) URI.
+ */
+export function resolveURI(uri: string): { repo: string; rev: string; path: string } {
+    const url = new URL(uri)
+    if (url.protocol === 'git:') {
+        return {
+            repo: (url.host + url.pathname).replace(/^\/*/, '').toLowerCase(),
+            rev: url.search.slice(1).toLowerCase(),
+            path: url.hash.slice(1),
+        }
+    }
+    throw new Error(`unrecognized URI: ${JSON.stringify(uri)} (supported URI schemes: git)`)
+}
+
+export interface Hunk {
+    startLine: number
+    endLine: number
+    author: {
+        person: {
+            email: string
+            displayName: string
+            user: {
+                username: string
+            } | null
+        }
+        date: string
+    }
+    rev: string
+    message: string
+    commit: {
+        url: string
+    }
+}
+
+export const queryBlameHunks = async (uri: string): Promise<Hunk[]> => {
+    const { repo, rev, path } = resolveURI(uri)
+    const { data, errors } = await sourcegraph.commands.executeCommand(
+        'queryGraphQL',
+        gql`
+            query GitBlame($repo: String!, $rev: String!, $path: String!) {
+                repository(name: $repo) {
+                    commit(rev: $rev) {
+                        blob(path: $path) {
+                            blame(startLine: 0, endLine: 0) {
```

Comment:
**question**: Are we querying all lines on each selection change? Should we either query all blame hunks on editor change, and call `getAuthorForLine` on selection changes, or query the blame hunks for the current selection on selection change?

Reply:
You're right! The current way is inefficient. I've now changed it so that it queries the blame only for the current line, when the selection changes. I chose to change it this way because it's a smaller change to make, although I recognize that your idea to query and cache all hunks on editor change would be a better further optimization.